File: src/main/java/AoC2024_03.java
import static java.lang.Integer.parseInt;
import static java.util.stream.Collectors.joining;

import java.util.List;
import java.util.regex.MatchResult;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2024_03
        extends SolutionBase<String, Integer, Integer> {
    
    private final Pattern REGEX = Pattern.compile
            ("(do(n't)?)\\(\\)|mul\\((\\d{1,3}),(\\d{1,3})\\)");

    private AoC2024_03(final boolean debug) {
        super(debug);
    }
    
    public static AoC2024_03 create() {
        return new AoC2024_03(false);
    }
    
    public static AoC2024_03 createDebug() {
        return new AoC2024_03(true);
    }
    
    @Override
    protected String parseInput(final List<String> inputs) {
        return inputs.stream().collect(joining());
    }
    
    private int solve(final String input, final boolean useConditionals) {
        int ans = 0;
        boolean enabled = true;
        final Matcher matcher = REGEX.matcher(input);
        while (matcher.find()) {
            final MatchResult m = matcher.toMatchResult();
            if (m.group(0).equals("do()")) {
                enabled = true;
            } else if (m.group(0).equals("don't()")) {
                enabled = false;
            } else {
                if (!useConditionals || enabled) {
                    ans += parseInt(m.group(3)) * parseInt(m.group(4));
                }
            }
        }
        return ans;
    }
    
    @Override
    public Integer solvePart1(final String input) {
        return solve(input, false);
    }
    
    @Override
    public Integer solvePart2(final String input) {
        return solve(input, true);
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "161"),
        @Sample(method = "part2", input = TEST2, expected = "48"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2024_03.create().run();
    }

    private static final String TEST1 = """
            xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))
            """;

    private static final String TEST2 = """
            xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))
            """;
}

File: src/main/python/AoC2024_03.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 3
#

import re
import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Input = str
Output1 = int
Output2 = int


TEST_1 = """\
xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))
"""
TEST_2 = """\
xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return "\n".join(line for line in input_data)

    def solve(self, memory: str, *, use_conditionals: bool) -> int:
        enabled = True
        ans = 0
        for do, _, a, b in re.findall(
            r"(do(n't)?)\(\)|mul\((\d{1,3}),(\d{1,3})\)", memory
        ):
            if do == "do":
                enabled = True
            elif do == "don't":
                enabled = False
            elif not use_conditionals or enabled:
                ans += int(a) * int(b)
        return ans

    def part_1(self, memory: Input) -> Output1:
        return self.solve(memory, use_conditionals=False)

    def part_2(self, memory: Input) -> Output2:
        return self.solve(memory, use_conditionals=True)

    @aoc_samples(
        (
            ("part_1", TEST_1, 161),
            ("part_2", TEST_2, 48),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 3)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
